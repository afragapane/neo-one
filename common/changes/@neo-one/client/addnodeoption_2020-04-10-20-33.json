{
  "changes": [
    {
      "packageName": "@neo-one/client",
      "comment": "Add NEODataProvider to support C# node.  Add NEOProvider to support NEODataProvider.  Moved common code to DataProviderBase & ProviderBase.",
      "type": "minor"
    }
  ],
  "packageName": "@neo-one/client",
  "email": "ahfragapane@gmail.com"
}