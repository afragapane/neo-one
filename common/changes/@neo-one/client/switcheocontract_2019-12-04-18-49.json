{
  "changes": [
    {
      "packageName": "@neo-one/client",
      "comment": "",
      "type": "none"
    }
  ],
  "packageName": "@neo-one/client",
  "email": "ahfragapane@gmail.com"
}