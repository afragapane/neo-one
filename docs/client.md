---
id: client
title: Client
---
This is the API documentation for the NEOâ€¢ONE Client.

## Properties

```
currentAccount$: Observable<UserAccount | undefined>
```
* Observable which returns a stream containing the currently active User Account.
<br><br>
```
accounts$: Observable<ReadonlyArray<UserAccount>>
```
* Observable which returns a stream containing an array of all User Accounts registered with the client.
<br><br>
```
currentNetwork$: Observable<NetworkType>
```
* Observable which returns a stream containing the currently active network.
<br><br>
```
networks$: Observable<ReadonlyArray<NetworkType>>
```
* Observable which returns a stream containing an array of all networks registered with the client.
<br><br>
```
block$: Observable<{block: Block, network: NetworkType}>
```
* Observable which returns a stream of blocks and their associated network as they are received.
<br><br>
```
accountState$: Observable<{currentAccount: UserAccount, account: Account} | undefined>
```
* Observable which returns a stream containing all information on the currently active account.
<br><br>
```
accountState$: Observable<{currentAccount: UserAccount, account: Account} | undefined>
```
* Observable which returns a stream containing all information on the currently active account.
```
providers: TUserAccountProviders
```
* Returns the User Account Providers.
<br><br>

```
hooks: ClientHooks
```
* Returns all client hooks.
<br><br><br>
## Methods
```
getAccount(idIn: UserAccountID): UserAccount
```
* Returns the UserAccount for a given UserAccountID.
<br><br>
```
selectAccount(idIn?: UserAccountID): Promise<void>
```
* Sets the current account to the account associated with the given UserAccountID.
<br><br>
```
selectNetwork(networkIn: NetworkType): Promise<void>
```
* Sets the currently selected network to given network.
<br><br>
```
deleteAccount(idIn: UserAccountID): Promise<void>
```
* Removes the account associated with the given UserAccountID from the list of accounts registered with the client.
<br><br>
```
updateAccountName(options: UpdateAccountNameOptions): Promise<void>
```
* Updates the name associated with a given account
<br><br>
```
getCurrentAccount(): UserAccount | undefined
```
* Returns the currently selected User Account. Returns undefined if no account is selected.
<br><br>
```
getCurrentNetwork(): NetworkType
```
* Returns the currently selected network.
<br><br>
```
getAccounts(): ReadonlyArray<UserAccount>
```
* Returns an array of all accounts registered with the client.
<br><br>
```
getNetworks(): ReadonlyArray<NetworkType>
```
* Returns an array of all networks registered with the client.
<br><br>
```
getAccounts(): ReadonlyArray<UserAccount>
```
* Returns an array of all accounts registered with the client.
<br><br>
```
transfer(amount: BigNumber, asset: Hash256String, to: AddressString, options?: TransactionOptions): Promise<TransactionResult<TransactionReceipt, ContractTransaction>>

transfer(transfers: ReadonlyArray<Transfer>, options?: TransactionOptions): Promise<TransactionResult>
```
* Transfer a given asset to a given address in the amount specified.
* Transfer will be from the currently selected account unless a from account is specified in the options.
* A network or system fee may also be included in the options.
* Transfer in batch by passing in an array of Transfers.
<br><br>
```
publish(contract: ContractRegister, options?: TransactionOptions): Promise<TransactionResult<PublishReceipt, InvocationTransaction>>
```
* Publishes a given smart contract to the currently selected network.
<br><br>
```
publishAndDeploy(contract: ContractRegister, abi: ABI, params: ReadonlyArray<Param> = [], options?: TransactionOptions, sourceMaps: Promise<SourceMaps> = Promise.resolve({})): Promise<TransactionResult<PublishReceipt, InvocationTransaction>>
```
* Publishes and deploys a smart contract to the currently selected network.
<br><br>
```
registerAsset(asset: AssetRegister, options?: TransactionOptions): Promise<TransactionResult<RegisterAssetReceipt, InvocationTransaction>>
```
* Registers a new asset to the currently selected network.
<br><br>
```
issue(amount: BigNumber, asset: Hash256String, to: AddressString, options?: TransactionOptions): Promise<TransactionResult<TransactionReceipt, IssueTransaction>>
```
* Issues a given amount of a given asset to a given address.
<br><br>
```
read(network: NetworkType): ReadClient
```
* Returns a ReadClient for the given network.  The ReadClient is used to obtain information about the blockchain.
<br><br>
```
smartContract<T extends SmartContract>(definition: SmartContractDefinition): T
```
* Creates a smart contract from a given smart contract definition.
<br><br>
```
inject<K>(provider: K extends keyof TUserAccountProviders ? TUserAccountProviders[K]: UserAccountProvider): void
```
* Sets the current User Account Provider to the given provider.
<br><br>
