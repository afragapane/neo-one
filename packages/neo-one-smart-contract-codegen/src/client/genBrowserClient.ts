import { NetworkDefinition, Wallet } from './types';

export const genBrowserClient = ({
  localDevNetworkName,
  wallets,
  networks,
}: {
  readonly localDevNetworkName: string;
  readonly wallets: ReadonlyArray<Wallet>;
  readonly networks: ReadonlyArray<NetworkDefinition>;
}) => ({
  js: `
import {
  Client,
  DeveloperClient,
  LocalKeyStore,
  LocalMemoryStore,
  LocalUserAccountProvider,
  NEOONEProvider,
  NEOONEDataProvider,
} from '@neo-one/client';
import { getJSONRPCLocalProviderManager } from '@neo-one/local-singleton';

const getDefaultUserAccountProviders = (provider) => ({
  memory: new LocalUserAccountProvider({
    keystore: new LocalKeyStore(new LocalMemoryStore()),
    provider,
  }),
});

const isLocalUserAccountProvider = (userAccountProvider) =>
  userAccountProvider instanceof LocalUserAccountProvider;

export const createClient = (getUserAccountProviders = getDefaultUserAccountProviders) => {
  const providers = [
    ${networks
      .filter(({ name }) => name !== localDevNetworkName)
      .map(({ name, rpcURL }) => `{ network: '${name}', rpcURL: '${rpcURL}' },`)
      .join('\n    ')}
  ];
  if (process.env.NODE_ENV !== 'production' || process.env.NEO_ONE_DEV === 'true') {
    providers.push(new NEOONEDataProvider({ network: '${localDevNetworkName}', rpcURL: getJSONRPCLocalProviderManager() }));
  }
  const provider = new NEOONEProvider(providers);
  const userAccountProviders = getUserAccountProviders(provider);
  const localUserAccountProviders = Object.values(userAccountProviders).filter(isLocalUserAccountProvider);
  const localUserAccountProvider = localUserAccountProviders.find(
    (userAccountProvider) => userAccountProvider.keystore instanceof LocalKeyStore,
  );
  if (process.env.NODE_ENV !== 'production' || process.env.NEO_ONE_DEV === 'true') {
    if (localUserAccountProvider !== undefined) {
      const localKeyStore = localUserAccountProvider.keystore;
      if (localKeyStore instanceof LocalKeyStore) {
        Promise.all([
          ${wallets
            .map(
              ({ name, wif }) =>
                `localKeyStore.addUserAccount({ network: '${localDevNetworkName}', name: '${name}', privateKey: '${wif}' }),`,
            )
            .join('\n          ')}
        ]).catch(() => {
          // do nothing
        });
      }
    }
  }

  return new Client(userAccountProviders);
};

export const createDeveloperClients = () => ({${networks
    .filter(({ name, dev }) => dev && name !== localDevNetworkName)
    .map(
      ({ name, rpcURL }) =>
        `\n  '${name}': new DeveloperClient(new NEOONEDataProvider({ network: '${name}', rpcURL: '${rpcURL}' })),`,
    )
    .join('')}
  '${localDevNetworkName}': new DeveloperClient(new NEOONEDataProvider({ network: '${localDevNetworkName}', rpcURL: getJSONRPCLocalProviderManager() })),
});
  `,
  ts: `
import {
  DeveloperClients,
  LocalKeyStore,
  LocalUserAccountProvider,
  NEOONEProvider,
  UserAccountProviders,
} from '@neo-one/client';

export interface DefaultUserAccountProviders {
  readonly memory: LocalUserAccountProvider<LocalKeyStore, NEOONEProvider>,
}

export const createClient: <TUserAccountProviders extends UserAccountProviders<any> = DefaultUserAccountProviders>(
  getUserAccountProviders?: (provider: NEOONEProvider) => TUserAccountProviders,
) => Client<TUserAccountProviders extends UserAccountProviders<infer TUserAccountProvider> ? TUserAccountProvider : never, TUserAccountProviders>;

export const createDeveloperClients: () => DeveloperClients;
`,
});
