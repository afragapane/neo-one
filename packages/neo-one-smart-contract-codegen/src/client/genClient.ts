export interface NetworkDefinition {
  readonly name: string;
  readonly rpcURL: string;
}

export const genClient = ({
  devNetworkName,
  masterPrivateKey,
  networks,
}: {
  readonly devNetworkName: string;
  readonly masterPrivateKey: string;
  readonly networks: ReadonlyArray<NetworkDefinition>;
}): string =>
  `
import { Client, LocalKeyStore, LocalMemoryStore, LocalUserAccountProvider, NEOONEProvider, UserAccountProvider } from '@neo-one/client';

export type DefaultUserAccountProviders = {
  readonly memory: LocalUserAccountProvider<LocalKeyStore, NEOONEProvider>,
}
const getDefaultUserAccountProviders = (provider: NEOONEProvider): DefaultUserAccountProviders => ({
  memory: new LocalUserAccountProvider({
    keystore: new LocalKeyStore({ store: new LocalMemoryStore() }),
    provider,
  }),
});

const isLocalUserAccountProvider = (userAccountProvider: UserAccountProvider): userAccountProvider is LocalUserAccountProvider<any, any> =>
  userAccountProvider instanceof LocalUserAccountProvider;

export const createClient = <TUserAccountProviders extends { readonly [K: string]: UserAccountProvider } = DefaultUserAccountProviders>(
  getUserAccountProviders: (provider: NEOONEProvider) => TUserAccountProviders = getDefaultUserAccountProviders as any,
): Client<TUserAccountProviders> => {
  const provider = new NEOONEProvider([
    ${networks.map(({ name, rpcURL }) => `{ network: '${name}', rpcURL: '${rpcURL}' },`).join('\n    ')}
  ]);
  const userAccountProviders = getUserAccountProviders(provider);

  const localUserAccountProviders = Object.values(userAccountProviders).filter(isLocalUserAccountProvider);
  const localUserAccountProvider = localUserAccountProviders.find(
    (userAccountProvider) => userAccountProvider.keystore instanceof LocalKeyStore,
  );
  if (localUserAccountProvider !== undefined) {
    const localKeyStore = localUserAccountProvider.keystore;
    if (localKeyStore instanceof LocalKeyStore) {
      localKeyStore.addAccount({
        network: '${devNetworkName}',
        name: 'master',
        privateKey: '${masterPrivateKey}',
      }).catch(() => {
        // do nothing
      });
    }
  }

  return new Client(userAccountProviders);
}
`;
