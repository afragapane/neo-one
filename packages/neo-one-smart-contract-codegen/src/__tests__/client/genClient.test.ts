import { genClient } from '../../client';

describe('genClient', () => {
  test('Token', () => {
    expect(
      genClient({
        localDevNetworkName: 'local',
        wallets: [
          {
            name: 'master',
            privateKey: 'L4qhHtwbiAMu1nrSmsTP5a3dJbxA3SNS6oheKnKd8E7KTJyCLcUv',
          },
        ],
        networks: [
          {
            name: 'local',
            rpcURL: 'http://localhost:4500/rpc',
            dev: true,
          },
        ],
        projectIDPath: '/foo/bar/one/generated/projectID.js',
        clientPath: '/foo/bar/one/generated/client.js',
        httpServerPort: 40100,
        browser: false,
      }),
    ).toMatchSnapshot();
  });
});
