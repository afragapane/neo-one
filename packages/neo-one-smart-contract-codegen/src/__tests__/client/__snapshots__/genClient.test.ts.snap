// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genClient Token 1`] = `
"
import { Client, LocalKeyStore, LocalMemoryStore, LocalUserAccountProvider, NEOONEProvider, UserAccountProvider } from '@neo-one/client';

export type DefaultUserAccountProviders = {
  readonly memory: LocalUserAccountProvider<LocalKeyStore, NEOONEProvider>,
}
const getDefaultUserAccountProviders = (provider: NEOONEProvider): DefaultUserAccountProviders => ({
  memory: new LocalUserAccountProvider({
    keystore: new LocalKeyStore({ store: new LocalMemoryStore() }),
    provider,
  }),
});

const isLocalUserAccountProvider = (userAccountProvider: UserAccountProvider): userAccountProvider is LocalUserAccountProvider<any, any> =>
  userAccountProvider instanceof LocalUserAccountProvider;

export const createClient = <TUserAccountProviders extends { readonly [K: string]: UserAccountProvider } = DefaultUserAccountProviders>(
  getUserAccountProviders: (provider: NEOONEProvider) => TUserAccountProviders = getDefaultUserAccountProviders as any,
): Client<TUserAccountProviders> => {
  const provider = new NEOONEProvider([
    { network: 'local', rpcURL: 'http://localhost:4500/rpc' },
  ]);
  const userAccountProviders = getUserAccountProviders(provider);

  const localUserAccountProviders = Object.values(userAccountProviders).filter(isLocalUserAccountProvider);
  const localUserAccountProvider = localUserAccountProviders.find(
    (userAccountProvider) => userAccountProvider.keystore instanceof LocalKeyStore,
  );
  if (localUserAccountProvider !== undefined) {
    const localKeyStore = localUserAccountProvider.keystore;
    if (localKeyStore instanceof LocalKeyStore) {
      localKeyStore.addAccount({
        network: 'local',
        name: 'master',
        privateKey: 'L4qhHtwbiAMu1nrSmsTP5a3dJbxA3SNS6oheKnKd8E7KTJyCLcUv',
      }).catch(() => {
        // do nothing
      });
    }
  }

  return new Client(userAccountProviders);
}
"
`;
